### DISCUSSION TITLE
Is it possible to dynamically create targets from a rule?

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#2770] [Q&A] [answered]
I have a target kind that needs to depend on another target, with custom configuration.  I'm currently adding it as a dependency and then modifying it in `after_load` in the rule, but this means that I can only have one instance of something that uses my rule, because otherwise they will both modify the target of the dependency in conflicting ways.

Is it possible to clone a target from a rule and then add the clone as a dependency?  Or explicitly construct a new target?  The constructor for `target` appears to need a `targetinfo` and I'm not sure how to create one of those.

**Comment by waruqi** [comment, marked as answer]
Clone target in rule is not supported and it is not recommended to modify the configuration of dependent targets in sub-targets.

**Reply by davidchisnall** [reply]
Is there another way of accomplishing the same thing?

**Reply by waruqi** [reply]
but I still don't understand what you're want to do.

**Reply by waruqi** [reply]
Why does a custom target kind depend on target?

**Reply by davidchisnall** [reply]
We are building a firmware image for an RTOS.  The final target for the image needs to depend on the scheduler but, because it's for an embedded system, we want to configure some of our components based on the configuration and code that the user provides.  My current approach has:

 - A target for each component that we're linking, including RTOS components and user-provided code.
 - A target for the final firmware image.
 - A rule to build a component and provide a partially linked output, imported by the component targets.
 - A rule to build the final firmware image, imported by the firmware target.

Users should not be exposed to the details of how the RTOS-provided components are built, but the scheduler (and some other components) need to be built in a configuration that depends on the properties of the firmware image target, including the set of components that the user added and the number of threads that they want to create.  

The current approach is to provide targets for each of the components that we provide and, in the rule that the user adopts for the firmware image, modify these targets to configure them according to the build.

In the CMake build system that we're currently using (and which I was evaluating xmake to replace), we have an `add_firmware` function that creates new targets for our code, configured as the user requires.

**Reply by waruqi** [reply]
Is it possible to create all the component targets in advance of the description scope? then we can select them in image rule.

```lua
target("component1")
    add_rules("build_component1")

target("component2")
    add_rules("build_component2")

target("image")
    on_load(function (target)
         if ... then
              target:add("deps", "component1")
         else
              target:add("deps", "component2")
         end
    end)
```

**Reply by waruqi** [reply]
We can also dynamically create targets in the description scope.

```lua
function add_firmware(name)
    target("component_" .. name)
        add_rules("build_component_" .. name)
end

add_firmware("xxx")
add_firmware("yyy")
```

**Reply by davidchisnall** [reply]
I hadn't managed to write functions in the description scope yet, but that might be feasible.  Is there a way of running code at the end?  Ideally I'd like adding firmware to look as much like adding any other kind of target, so someone could do something like:

```
firmware("example")
  add_files(...)
  add_cflags(...)
  ...
```

If I return a target that has uses the `after_load` hook, I could do the setup there, so my function would create the dependent targets and then configure them after the target that depends on them has been configured.  (More generally, does it leave the last target that I define open?  Do I need to explicitly return it somehow?)

**Reply by davidchisnall** [reply]
It looks as if this does work, thanks!  

Unfortunately, when I declare the function, the directory that `os.scriptdir()` returns is the directory of the script for the project, not the script for the file containing the function (which has a fixed path relative to the source files that I want to include).  Is there a simple way of getting the path of the file containing the currently executing function?  Or of capturing this when the file is loaded so that I can refer to it later (I couldn't work out how to create / initialise variables in the description scope)?

**Reply by waruqi** [reply]
you can use `os.projectdir()`, it is always project root directory.

>  Is there a simple way of getting the path of the file containing the currently executing function?

try 

```lua
local scriptdir = os.scriptdir()
function add_firmware()
    print(scriptdir)
end
```

**Reply by davidchisnall** [reply]
I am encountering problems with this and I suspect that the root cause is the same as some of my other issues.

The xmake.lua that provided the `firmware` description-scope function was in the root of my repo, for historical reasons.  It's now been refactored so that it's in the `sdk` directory and now suddenly the `firmware` function is not visible to other things.  I think the implicit scoping from having it in the root was the problem.

How do I provide a set of description-scope functions that I can `include` (or `import`) into a file that is not in a subdirectory of them?

**Comment by waruqi** [comment]
I am supporting create target/rule dynamically in script scope. https://github.com/xmake-io/xmake/issues/2879

maybe you can try it later, but it has been not merged yet. I am still developing it.
